crates/paths/src/paths.rs

```diff
@@ -5,61 +5,84 @@ use std::sync::OnceLock;
 
 pub use util::paths::home_dir;
 
+/// A default editorconfig file name to use when resolving project settings.
+pub const EDITORCONFIG_NAME: &str = ".editorconfig";
+
+// Custom data directory override, set only by set_custom_data_dir.
+static CUSTOM_DATA_DIR: OnceLock<PathBuf> = OnceLock::new();
+
+// Resolved data directory, combining custom or platform defaults, set once.
+static CURRENT_DATA_DIR: OnceLock<PathBuf> = OnceLock::new();
+
+// Resolved config directory, combining custom or platform defaults, set once.
+static CONFIG_DIR: OnceLock<PathBuf> = OnceLock::new();
+
 /// Returns the relative path to the zed_server directory on the ssh host.
 pub fn remote_server_dir_relative() -> &'static Path {
     Path::new(".zed_server")
 }
 
+/// Sets a custom directory for all user data, overriding the default data directory.
+///
+/// Panics if called after the data directory has been initialized (e.g., via `data_dir` or `config_dir`).
+pub fn set_custom_data_dir(dir: &str) -> &'static PathBuf {
+    if CURRENT_DATA_DIR.get().is_some() || CONFIG_DIR.get().is_some() {
+        panic!("set_custom_data_dir called after data_dir or config_dir was initialized");
+    }
+    CUSTOM_DATA_DIR.get_or_init(|| {
+        let path = PathBuf::from(dir);
+        std::fs::create_dir_all(&path).expect("failed to create custom data directory");
+        path
+    })
+}
+
 /// Returns the path to the configuration directory used by Zed.
 pub fn config_dir() -> &'static PathBuf {
-    static CONFIG_DIR: OnceLock<PathBuf> = OnceLock::new();
```

Comment:
I don't think there's any reason to move these out of their functions?

Reply:
I moved `CONFIG_DIR` and `CURRENT_DATA_DIR` out of the functions so `set_custom_data_dir` can check their initialization state and panic if they’ve already been set. Keeping them inside would prevent this check. They’re still private to the module (not `pub`), so it doesn’t affect external access. I could refactor to avoid this (e.g., by tracking initialization differently), but I think this approach is cleanest for now. Happy to adjust if you have a better suggestion!

Reply:
Ahh, I see. Thank you!